staging/src/k8s.io/api/networking/v1/types.go

```diff
@@ -201,3 +201,300 @@ type NetworkPolicyList struct {
+	//   present in an Ingress spec, the longest matching path (/foo/) is given
+	//   priority.
+	PathTypePrefix = PathType("Prefix")
+
+	// PathTypeImplementationSpecific matching is up to the IngressClass.
+	// Implementations can treat this as a separate PathType or treat it
+	// identically to Prefix or Exact path types.
+	PathTypeImplementationSpecific = PathType("ImplementationSpecific")
+)
+
+// HTTPIngressPath associates a path with a backend. Incoming urls matching the
+// path are forwarded to the backend.
+type HTTPIngressPath struct {
+	// Path is matched against the path of an incoming request. Currently it can
+	// contain characters disallowed from the conventional "path" part of a URL
+	// as defined by RFC 3986. Paths must begin with a '/'. When unspecified,
+	// all paths from incoming requests are matched.
+	// +optional
+	Path string `json:"path,omitempty" protobuf:"bytes,1,opt,name=path"`
+
+	// PathType determines the interpretation of the Path matching. PathType can
+	// be one of the following values:
+	// * Exact: Matches the URL path exactly.
+	// * Prefix: Matches based on a URL path prefix split by '/'. Matching is
+	//   done on a path element by element basis. A path element refers is the
+	//   list of labels in the path split by the '/' separator. A request is a
+	//   match for path p if every p is an element-wise prefix of p of the
+	//   request path. Note that if the last element of the path is a substring
+	//   of the last element in request path, it is not a match (e.g. /foo/bar
+	//   matches /foo/bar/baz, but does not match /foo/barbaz).
+	// * ImplementationSpecific: Interpretation of the Path matching is up to
+	//   the IngressClass. Implementations can treat this as a separate PathType
+	//   or treat it identically to Prefix or Exact path types.
+	// Implementations are required to support all path types.
+	// Defaults to ImplementationSpecific.
+	PathType *PathType `json:"pathType,omitempty" protobuf:"bytes,3,opt,name=pathType"`
+
+	// Backend defines the referenced service endpoint to which the traffic
+	// will be forwarded to.
+	Backend IngressBackend `json:"backend" protobuf:"bytes,2,opt,name=backend"`
+}
+
+// IngressBackend describes all endpoints for a given service and port.
+type IngressBackend struct {
+	// Specifies the name of the referenced service.
+	// +optional
+	ServiceName string `json:"serviceName,omitempty" protobuf:"bytes,1,opt,name=serviceName"`
+
+	// Specifies the port of the referenced service.
+	// +optional
+	ServicePort intstr.IntOrString `json:"servicePort,omitempty" protobuf:"bytes,2,opt,name=servicePort"`
+
+	// Resource is an ObjectRef to another Kubernetes resource in the namespace
+	// of the Ingress object. If resource is specified, serviceName and servicePort
+	// must not be specified.
+	// +optional
+	Resource *v1.TypedLocalObjectReference `json:"resource,omitempty" protobuf:"bytes,3,opt,name=resource"`
```

Comment:
this is missing the restructuring from https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190125-ingress-api-group.md#alternative-backend-types:

```
type IngressBackend struct {
  Service *ServiceBackend
  Resource *v1.TypedLocalObjectReference
}
```

this will mean:
* updating the internal type structure
* defining conversions for extensions/v1beta1 and networking/v1beta1 to round-trip the service name/port fields
* updating validation to make the field paths for servicePort/serviceName version-specific

Reply:
Was going to do this in a follow-up PR but it makes sense to do no

Reply:
For the old field, do I need to preserve any field numbers or names when I move to the new restructuring?

i.e. Do I need to keep these json tag names

```
        // Specifies the name of the referenced service.
	// +optional
	ServiceName string `json:"serviceName,omitempty" protobuf:"bytes,1,opt,name=serviceName"`

	// Specifies the port of the referenced service.
	// +optional
	ServicePort intstr.IntOrString `json:"servicePort,omitempty" protobuf:"bytes,2,opt,name=servicePort"`
```

Reply:
no, v1 serialization is independent. I'd expect something like:

```go
type IngressBackend struct {
	Service *IngressServiceBackend `json:"service,omitempty" ...`
	Resource *v1.TypedLocalObjectReference `json:"resource,omitempty" ...`
}

type IngressServiceBackend struct {
	Name string `json:"name,omitempty" ...`
	ServicePort intstr.IntOrString `json:"port,omitempty" ...`
}
```

proto generation will add proto tags and tag numbers for you.

Reply:
So far I have the following

```
// IngressBackend describes all endpoints for a given service and port.
type IngressBackend struct {
	// Only one of the following fields may be specified.

	// Service references a Service as a Backend. This is specially
	// called out as it is required to be supported AND to reduce
	// verbosity.
	// +optional
	Service *ServiceBackend `json:"serviceBackend,omitempty" protobuf:"bytes,4,opt,name=serviceBackend"`

	// Resource is an ObjectRef to another Kubernetes resource in the namespace
	// of the Ingress object. If resource is specified, serviceName and servicePort
	// must not be specified.
	// +optional
	Resource *v1.TypedLocalObjectReference `json:"resource,omitempty" protobuf:"bytes,3,opt,name=resource"`
}

// ServiceBackend references a Kubernetes Service as a Backend.
type ServiceBackend struct {
	// Name is the referenced service. The service must exist in
	// the same namespace as the Ingress object.
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`

	// Port of the referenced service. If unspecified and the ServiceName is
	// non-empty, the Service must expose a single port.
	// +optional
	Port ServiceBackendPort `json:"port" protobuf:"bytes,2,opt,name=port"`
}

// ServiceBackendPort is the service port being referenced.
type ServiceBackendPort struct {
	// Name is the name of the port on the Service.
	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
	// Number is the numerical port number (e.g. 80) on the Service.
	Number int `json:"number" protobuf:"bytes,2,opt,name=number"`	
}
```

Reply:
* s/serviceBackend/service/
* I assume name is required if a service backend is specified, so drop the +optional / omitempty bits
* I assume name and number are mutually exclusive in ServiceBackendPort, so mark them as +optional / omitempty

Reply:
@liggitt Do you think that the split ServiceBackendPort makes sense over the existing intstr.IntOrString approach? If so should ServiecBackendPort.Number be a required field like it is in nonheadless core.Service.Spec.Ports.Port ?

Reply:
cc @thockin on the above comment ^

Reply:
with the split port name/number fields, what would an existing v1beta1 Ingress object that had a string `servicePort` (like `servicePort: "http"`) convert to in v1? it seems like it would have to convert to `service: {port: {name: "http"}}`. If that's the case, then port.number can't be required.

Reply:
I'm generally in favor of explicit separate number and string fields... intOrString has not played nicely with openapi or generated clients in the past